pkg/controller/workload/job/job_controller.go

```diff
@@ -210,39 +207,41 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 	log := ctrl.LoggerFrom(ctx).WithValues("job", klog.KObj(&job))
 	ctx = ctrl.LoggerInto(ctx, log)
 
-	pwName := parentWorkload(&job)
+	isMainJob := parentWorkloadName(&job) == ""
 
 	// when manageJobsWithoutQueueName is disabled we only reconcile jobs that have either
 	// queue-name or the parent-workload annotation set.
-	if queueName(&job) == "" && pwName == "" && !r.manageJobsWithoutQueueName {
+	if !r.manageJobsWithoutQueueName && queueName(&job) == "" && isMainJob {
 		log.V(3).Info(fmt.Sprintf("Neither %s, nor %s annotation is set, ignoring the job", constants.QueueAnnotation, constants.ParentWorkloadAnnotation))
 		return ctrl.Result{}, nil
 	}
 
 	log.V(2).Info("Reconciling Job")
 
-	var childWorkloads kueue.WorkloadList
-	if err := r.client.List(ctx, &childWorkloads, client.InNamespace(req.Namespace),
-		client.MatchingFields{ownerKey: req.Name}); err != nil {
-		log.Error(err, "Unable to list child workloads")
-		return ctrl.Result{}, err
-	}
-
 	// 1. make sure there is only a single existing instance of the workload
-	wl, err := r.ensureAtMostOneWorkload(ctx, &job, childWorkloads)
+	wl, err := r.ensureAtMostOneWorkload(ctx, &job)
 	if err != nil {
 		log.Error(err, "Getting existing workloads")
 		return ctrl.Result{}, err
 	}
 
-	jobFinishedCond, jobFinished := jobFinishedCondition(&job)
-	// 2. create new workload if none exists
-	if wl == nil {
-		// Nothing to do if the job is finished
-		if jobFinished {
+	// 2. handle job is finished.
+	if jobFinishedCond, jobFinished := jobFinishedCondition(&job); jobFinished {
+		if !isMainJob || wl == nil || apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadFinished) {
 			return ctrl.Result{}, nil
 		}
-		if pwName != "" {
+		condition := generateFinishedCondition(jobFinishedCond)
+		apimeta.SetStatusCondition(&wl.Status.Conditions, condition)
+		err := r.client.Status().Update(ctx, wl)
+		if err != nil {
+			log.Error(err, "Updating workload status")
+		}
+		return ctrl.Result{}, err
+	}
+
+	// 3. handle workload is nil.
+	if wl == nil {
+		if !isMainJob {
 			return ctrl.Result{}, nil
```

Comment:
in this case, shouldn't we should suspend the dependent job?

so maybe it should be `if wl == nil && isStandaloneJob` and below (step 6) it should say `if wl == nil || wl.Spec.Admission == nil`

Reply:
We only create workload here, leaving the suspend logics to another  reconciler.
After creating the workload:
1. if dependent job is suspended, jump to step5.
2. if dependent job is unsuspended, jump to step6.

Reply:
But one corner case is both parent and child jobs are unsuspended, then we delete the workload manually. Then if the workload:
- admitted immediately again with the same assignment: it works well
- admitted immediately again with a different assignment: the node affinity may be changed so not working as expected
- unadmitted, jobs will be suspended finally.

Then we may have a potential bug here. We can `suspend all the jobs here when workload is nil but we should make another call` or  `validate the nodeSelector in "jobAndWorkloadEqual" also` .

Reply:
it could also be the case that somehow the parent job controller is crashing (and not able to create a workload) and somehow a user set the child job to suspend=false. We don't want that child job to continue running.

Reply:
open a ticket here https://github.com/kubernetes-sigs/kueue/issues/595

Reply:
@kerthcet can you double check this is an issue currently, we have this integration test: https://github.com/kubernetes-sigs/kueue/blob/b2a5e386d7e9c0e3346660dd01001734f631d7fd/test/integration/controller/job/job_controller_test.go#L245. The code to ensure that is here: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/workload/job/job_controller.go#L483-L495

Reply:
Sure, I'll take charge of this.

Reply:
I agree. This shouldn't be a problem in the existing code.

Reply:
Yes, I think it will not happen, I just complicated the situation about `admitted immediately again with a different assignment: the node affinity may be changed so not working as expected`, the admitted workload should be create at first, then at that moment, the job will be suspended.

Reply:
> it could also be the case that somehow the parent job controller is crashing (and not able to create a workload) and somehow a user set the child job to suspend=false. We don't want that child job to continue running.

This will also be covered by https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/workload/job/job_controller.go#L483-L495